### PR TITLE
Add eth_sign

### LINKED ISSUE #766: Eth RPC for eth_sign
Implement the eth_sign RPC in web3signer.

https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_sign

### PR DESCRIPTION
**Pull request by gfukushima**
## PR Description
This PR add the eth_sign RPC method to web3signer

## Fixed Issue(s)

Fixes #766

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  public EthSignResultProvider(final ArtifactSignerProvider transactionSignerProvider) {
+    this.transactionSignerProvider = transactionSignerProvider;
+  }
+
+  @Override
+  public String createResponseResult(final JsonRpcRequest request) {
+    final List<String> params = getParams(request);
+    if (params == null || params.size() != 2) {
+      LOG.info(
```
Comment (usmansaleem) [inline core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java:49]: I reckon this should be debug ... as we are already returning error/exception to the caller.

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          "eth_sign should have a list of 2 parameters, but has {}",
+          params == null ? "null" : params.size());
+      throw new JsonRpcException(INVALID_PARAMS);
+    }
+
+    final String address = params.get(0);
+    final Optional<ArtifactSigner> transactionSigner = transactionSignerProvider.getSigner(address);
+    if (transactionSigner.isEmpty()) {
+      LOG.info("Address ({}) does not match any available account", address);
```
Comment (usmansaleem) [inline core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java:58]: same, should this log be at debug level?

**Review by usmansaleem — APPROVED** [review]
LGTM - my only comment is around changing level of logs in EthSignResultProvider. Get opinion from other team members around it as well.

File: core/src/main/java/tech/pegasys/web3signer/core/util/EthMessageUtil.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.util.Locale;
+
+import org.apache.tuweni.bytes.Bytes;
+
+// adapted from
+// https://github.com/web3j/web3j/blob/7054ab324cab0d44a759780e61ee12978fd17490/crypto/src/main/java/org/web3j/crypto/Sign.java#L59
+public class EthMessageUtil {
```
Comment (jframe) [inline core/src/main/java/tech/pegasys/web3signer/core/util/EthMessageUtil.java:23]: Can you also bring across the corresponding test `EthMessageUtilTest`

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java (hunk 1 above)
Comment (jframe) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java:49]: Agree debug would be good

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java (hunk 2 above)
Comment (jframe) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java:58]: Same, think debug would be good here too

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    return secpSigner.sign(address, ethMessage).get();
+  }
+
+  private List<String> getParams(final JsonRpcRequest request) {
+    try {
+      @SuppressWarnings("unchecked")
+      final List<String> params = (List<String>) request.getParams();
+      return params;
+    } catch (final ClassCastException e) {
+      LOG.info(
```
Comment (jframe) [inline core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java:76]: Also debug here

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    final String address = params.get(0);
+    final Optional<ArtifactSigner> transactionSigner = transactionSignerProvider.getSigner(address);
+    if (transactionSigner.isEmpty()) {
+      LOG.info("Address ({}) does not match any available account", address);
+      throw new JsonRpcException(SIGNING_FROM_IS_NOT_AN_UNLOCKED_ACCOUNT);
+    }
+
+    final Bytes ethMessage = getEthereumMessage(params.get(1));
+
+    final SignerForIdentifier<SecpArtifactSignature> secpSigner =
```
Comment (jframe) [inline core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java:64]: Be good to pass in the SignerForIdentifier like what are doing in `Eth1SignForIdentifierHandler` so we don't need to construct this each time

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        new SignerForIdentifier<>(transactionSignerProvider, this::formatSecpSignature, SECP256K1);
+
+    return secpSigner.sign(address, ethMessage).get();
```
Comment (jframe) [inline core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java:67]: We should normalise the identifier to lowercase using IdentifierUtils.normaliseIdentifier like what is done in `Eth1SignForIdentifierHandler`.

 ```suggestion
    return secpSigner.sign(normaliseIdentifier(identifier), ethMessage).get();
```

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020 ConsenSys AG.
```
Comment (siladu) [inline core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java:2]: nit: 2023

```suggestion
 * Copyright 2023 ConsenSys AG.
```

File: core/src/main/java/tech/pegasys/web3signer/core/util/EthMessageUtil.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2021 ConsenSys AG.
```
Comment (siladu) [inline core/src/main/java/tech/pegasys/web3signer/core/util/EthMessageUtil.java:2]: nit: 2023

```suggestion
 * Copyright 2023 ConsenSys AG.
```

File: core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/EthSignResultProviderTest.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2020 ConsenSys AG.
```
Comment (siladu) [inline core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/EthSignResultProviderTest.java:2]: nit: 2023

```suggestion
 * Copyright 2023 ConsenSys AG.
```

**Review by siladu — COMMENTED** [review]
My first thought was the log messages should be WARN rather than DEBUG but will defer to the other guys who know the code better.

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java (hunk 2 above)
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java]: Done

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java (hunk 1 above)
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java]: done

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java (hunk 4 above)
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java]: Done

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java (hunk 5 above)
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java]: done

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java (hunk 6 above)
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java]: done

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java (hunk 7 above)
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/internalresponse/EthSignResultProvider.java]: done

File: core/src/main/java/tech/pegasys/web3signer/core/util/EthMessageUtil.java (hunk 8 above)
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/util/EthMessageUtil.java]: done

File: core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/EthSignResultProviderTest.java (hunk 9 above)
Comment (gfukushima) [reply core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/EthSignResultProviderTest.java]: done

**Comment by gfukushima** [comment]
Closing in favour of #807

**Pull request closed without merging**